File: repositories/src/main/java/ch/dissem/bitmessage/repository/JdbcNodeRegistry.java
/*
 * Copyright 2016 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.repository;

import ch.dissem.bitmessage.entity.valueobject.NetworkAddress;
import ch.dissem.bitmessage.exception.ApplicationException;
import ch.dissem.bitmessage.ports.NodeRegistry;
import ch.dissem.bitmessage.utils.Collections;
import ch.dissem.bitmessage.utils.SqlStrings;
import ch.dissem.bitmessage.utils.Strings;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static ch.dissem.bitmessage.ports.NodeRegistryHelper.loadStableNodes;
import static ch.dissem.bitmessage.utils.UnixTime.*;

public class JdbcNodeRegistry extends JdbcHelper implements NodeRegistry {
    private static final Logger LOG = LoggerFactory.getLogger(JdbcNodeRegistry.class);
    private Map<Long, Set<NetworkAddress>> stableNodes;

    public JdbcNodeRegistry(JdbcConfig config) {
        super(config);
        cleanUp();
    }

    private void cleanUp() {
        try (
            Connection connection = config.getConnection();
            PreparedStatement ps = connection.prepareStatement(
                "DELETE FROM Node WHERE time<?")
        ) {
            ps.setLong(1, now() - 28 * DAY);
            ps.executeUpdate();
        } catch (SQLException e) {
            LOG.error(e.getMessage(), e);
        }
    }

    private NetworkAddress loadExisting(NetworkAddress node) {
        String query =
            "SELECT stream, address, port, services, time" +
                " FROM Node" +
                " WHERE stream = " + node.getStream() +
                "   AND address = X'" + Strings.hex(node.getIPv6()) + "'" +
                "   AND port = " + node.getPort();
        try (
            Connection connection = config.getConnection();
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery(query)
        ) {
            if (rs.next()) {
                return new NetworkAddress.Builder()
                    .stream(rs.getLong("stream"))
                    .ipv6(rs.getBytes("address"))
                    .port(rs.getInt("port"))
                    .services(rs.getLong("services"))
                    .time(rs.getLong("time"))
                    .build();
            } else {
                return null;
            }
        } catch (Exception e) {
            LOG.error(e.getMessage(), e);
            throw new ApplicationException(e);
        }
    }

    @Override
    public void clear() {
        try (
            Connection connection = config.getConnection();
            PreparedStatement ps = connection.prepareStatement(
                "DELETE FROM Node")
        ) {
            ps.executeUpdate();
        } catch (SQLException e) {
            LOG.error(e.getMessage(), e);
        }
    }

    @Override
    public List<NetworkAddress> getKnownAddresses(int limit, long... streams) {
        List<NetworkAddress> result = new LinkedList<>();
        String query =
            "SELECT stream, address, port, services, time" +
                " FROM Node WHERE stream IN (" + SqlStrings.join(streams) + ")" +
                " ORDER BY TIME DESC" +
                " LIMIT " + limit;
        try (
            Connection connection = config.getConnection();
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery(query)
        ) {
            while (rs.next()) {
                result.add(
                    new NetworkAddress.Builder()
                        .stream(rs.getLong("stream"))
                        .ipv6(rs.getBytes("address"))
                        .port(rs.getInt("port"))
                        .services(rs.getLong("services"))
                        .time(rs.getLong("time"))
                        .build()
                );
            }
        } catch (Exception e) {
            LOG.error(e.getMessage(), e);
            throw new ApplicationException(e);
        }
        if (result.isEmpty()) {
            synchronized (this) {
                if (stableNodes == null) {
                    stableNodes = loadStableNodes();
                }
            }
            for (long stream : streams) {
                Set<NetworkAddress> nodes = stableNodes.get(stream);
                if (nodes != null && !nodes.isEmpty()) {
                    result.add(Collections.selectRandom(nodes));
                }
            }
            if (result.isEmpty()) {
                // There might have been an error resolving domain names due to a missing internet exception.
                // Try to load the stable nodes again next time.
                stableNodes = null;
            }
        }
        return result;
    }

    @Override
    public void offerAddresses(List<NetworkAddress> nodes) {
        cleanUp();
        nodes.stream()
            .filter(node -> node.getTime() < now() + 2 * MINUTE && node.getTime() > now() - 28 * DAY)
            .forEach(node -> {
                synchronized (this) {
                    NetworkAddress existing = loadExisting(node);
                    if (existing == null) {
                        insert(node);
                    } else if (node.getTime() > existing.getTime()) {
                        update(node);
                    }
                }
            });
    }

    private void insert(NetworkAddress node) {
        try (
            Connection connection = config.getConnection();
            PreparedStatement ps = connection.prepareStatement(
                "INSERT INTO Node (stream, address, port, services, time) " +
                    "VALUES (?, ?, ?, ?, ?)")
        ) {
            ps.setLong(1, node.getStream());
            ps.setBytes(2, node.getIPv6());
            ps.setInt(3, node.getPort());
            ps.setLong(4, node.getServices());
            ps.setLong(5, node.getTime());
            ps.executeUpdate();
        } catch (SQLException e) {
            LOG.error(e.getMessage(), e);
        }
    }

    private void update(NetworkAddress node) {
        try (
            Connection connection = config.getConnection();
            PreparedStatement ps = connection.prepareStatement(
                "UPDATE Node SET services=?, time=? WHERE stream=? AND address=? AND port=?")
        ) {
            ps.setLong(1, node.getServices());
            ps.setLong(2, node.getTime());
            ps.setLong(3, node.getStream());
            ps.setBytes(4, node.getIPv6());
            ps.setInt(5, node.getPort());
            ps.executeUpdate();
        } catch (SQLException e) {
            LOG.error(e.getMessage(), e);
        }
    }
}

File: repositories/src/main/kotlin/ch/dissem/bitmessage/repository/JdbcNodeRegistry.kt
/*
 * Copyright 2016 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.repository

import ch.dissem.bitmessage.entity.valueobject.NetworkAddress
import ch.dissem.bitmessage.ports.NodeRegistry
import ch.dissem.bitmessage.ports.NodeRegistryHelper.loadStableNodes
import ch.dissem.bitmessage.utils.*
import ch.dissem.bitmessage.utils.Collections
import ch.dissem.bitmessage.utils.UnixTime.DAY
import ch.dissem.bitmessage.utils.UnixTime.MINUTE
import ch.dissem.bitmessage.utils.UnixTime.now
import org.slf4j.LoggerFactory
import java.sql.SQLException
import java.util.*

class JdbcNodeRegistry(config: JdbcConfig) : JdbcHelper(config), NodeRegistry {
    private var stableNodes: Map<Long, Set<NetworkAddress>> = emptyMap()
        get() {
            if (field.isEmpty())
                field = loadStableNodes()
            return field
        }

    init {
        cleanUp()
    }

    private fun cleanUp() {
        try {
            config.getConnection().use { connection ->
                connection.prepareStatement("DELETE FROM Node WHERE time<?").use { ps ->
                    ps.setLong(1, now - 28 * DAY)
                    ps.executeUpdate()
                }
            }
        } catch (e: SQLException) {
            LOG.error(e.message, e)
        }
    }

    private fun loadExisting(node: NetworkAddress): NetworkAddress? {
        val query = """
            SELECT stream, address, port, services, time
            FROM Node
            WHERE stream = ${node.stream} AND address = X'${Strings.hex(node.IPv6)}' AND port = ${node.port}
        """
        config.getConnection().use { connection ->
            connection.createStatement().use { stmt ->
                stmt.executeQuery(query).use { rs ->
                    if (rs.next()) {
                        return NetworkAddress.Builder()
                            .stream(rs.getLong("stream"))
                            .ipv6(rs.getBytes("address"))
                            .port(rs.getInt("port"))
                            .services(rs.getLong("services"))
                            .time(rs.getLong("time"))
                            .build()
                    } else {
                        return null
                    }
                }
            }
        }
    }

    override fun clear() {
        try {
            config.getConnection().use { connection ->
                connection.prepareStatement("DELETE FROM Node").use { ps -> ps.executeUpdate() }
            }
        } catch (e: SQLException) {
            LOG.error(e.message, e)
        }
    }

    override fun getKnownAddresses(limit: Int, vararg streams: Long): List<NetworkAddress> {
        val result = LinkedList<NetworkAddress>()
        val query = """
            SELECT stream, address, port, services, time
            FROM Node
            WHERE stream IN (${SqlStrings.join(*streams)})
            ORDER BY TIME DESC LIMIT $limit
        """
        config.getConnection().use { connection ->
            connection.createStatement().use { stmt ->
                stmt.executeQuery(query).use { rs ->
                    while (rs.next()) {
                        result.add(
                            NetworkAddress.Builder()
                                .stream(rs.getLong("stream"))
                                .ipv6(rs.getBytes("address"))
                                .port(rs.getInt("port"))
                                .services(rs.getLong("services"))
                                .time(rs.getLong("time"))
                                .build()
                        )
                    }
                }
            }
        }

        if (result.isEmpty()) {
            streams
                .asSequence()
                .mapNotNull { stableNodes[it] }
                .filter { it.isNotEmpty() }
                .mapTo(result) { Collections.selectRandom(it) }
            if (result.isEmpty()) {
                // There might have been an error resolving domain names due to a missing internet connection.
                // Try to load the stable nodes again next time.
                stableNodes = emptyMap()
            }
        }
        return result
    }

    override fun offerAddresses(nodes: List<NetworkAddress>) {
        cleanUp()
        nodes.stream()
            .filter { (time) -> time < now + 2 * MINUTE && time > now - 28 * DAY }
            .forEach { node ->
                synchronized(this) {
                    val existing = loadExisting(node)
                    if (existing == null) {
                        insert(node)
                    } else if (node.time > existing.time) {
                        update(node)
                    }
                }
            }
    }

    private fun insert(node: NetworkAddress) {
        try {
            config.getConnection().use { connection ->
                connection.prepareStatement(
                    "INSERT INTO Node (stream, address, port, services, time) VALUES (?, ?, ?, ?, ?)").use { ps ->
                    ps.setLong(1, node.stream)
                    ps.setBytes(2, node.IPv6)
                    ps.setInt(3, node.port)
                    ps.setLong(4, node.services)
                    ps.setLong(5,
                        if (node.time > UnixTime.now) {
                            // This might be an attack, let's not use those nodes with priority
                            UnixTime.now - 7 * UnixTime.DAY
                        } else if (node.time == 0L) {
                            // Those just don't have a time set
                            // let's give them slightly higher priority than the possible attack ones
                            UnixTime.now - 6 * UnixTime.DAY
                        } else {
                            node.time
                        }
                    )
                    ps.executeUpdate()
                }
            }
        } catch (e: SQLException) {
            LOG.error(e.message, e)
        }
    }

    override fun update(node: NetworkAddress) {
        try {
            val time = if (node.time > UnixTime.now) {
                // This might be an attack, let's not use those nodes with priority
                UnixTime.now - 7 * UnixTime.DAY
            } else {
                node.time
            }

            config.getConnection().use { connection ->
                connection.prepareStatement(
                    "UPDATE Node SET services=?, time=? WHERE stream=? AND address=? AND port=?").use { ps ->
                    ps.setLong(1, node.services)
                    ps.setLong(2, max(node.time, time))
                    ps.setLong(3, node.stream)
                    ps.setBytes(4, node.IPv6)
                    ps.setInt(5, node.port)
                    ps.executeUpdate()
                }
            }
        } catch (e: SQLException) {
            LOG.error(e.message, e)
        }
    }

    override fun remove(node: NetworkAddress) {
        try {
            config.getConnection().use { connection ->
                connection.prepareStatement(
                    "DELETE FROM Node WHERE stream=? AND address=? AND port=?").use { ps ->
                    ps.setLong(1, node.stream)
                    ps.setBytes(2, node.IPv6)
                    ps.setInt(3, node.port)
                    ps.executeUpdate()
                }
            }
        } catch (e: SQLException) {
            LOG.error(e.message, e)
        }
    }

    override fun cleanup() {
        try {
            config.getConnection().use { connection ->
                connection.prepareStatement(
                    "DELETE FROM Node WHERE time<?").use { ps ->
                    ps.setLong(1, UnixTime.now - 8 * DAY)
                    ps.executeUpdate()
                }
            }
        } catch (e: SQLException) {
            LOG.error(e.message, e)
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(JdbcNodeRegistry::class.java)
    }
}
